In [ ]:
import torch
from torch import nn
from transformers.models.bert import BertModel,BertTokenizer,BertForMaskedLM

In [ ]:
model_type = 'bert-base-uncased'

In [ ]:
tokenizer = BertTokenizer.from_pretrained(model_type)
bert = BertModel.from_pretrained(model_type)
mlm = BertForMaskedLM.from_pretrained(model_type,output_hidden_states=True)

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

[transformers] loading configuration file config.json from cache at /root/.cache/huggingface/hub/models--bert-base-uncased/snapshots/86b5e0934494bd15c9632b12f734a8a67f723594/config.json
[transformers] Model config BertConfig {
  "add_cross_attention": false,
  "architectures": [
    "BertForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "bos_token_id": null,
  "classifier_dropout": null,
  "eos_token_id": null,
  "gradient_checkpointing": false,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "is_decoder": false,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "model_type": "bert",
  "num_attention_heads": 12,
  "num_hidden_layers": 12,
  "pad_token_id": 0,
  "position_embedding_type": "absolute",
  "tie_word_embeddings": true,
  "transformers_version": "5.12.1",
  "type_vocab_size": 2,
  "use_cache": true,
  "vocab_size": 30522
}



model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

[transformers] loading weights file model.safetensors from cache at /root/.cache/huggingface/hub/models--bert-base-uncased/snapshots/86b5e0934494bd15c9632b12f734a8a67f723594/model.safetensors
[transformers] Since the `dtype` attribute can't be found in model's config object, will use dtype=torch.float32 as derived from model's weights


Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

[transformers] BertModel LOAD REPORT from: bert-base-uncased
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
cls.predictions.transform.dense.weight     | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED |  | 
cls.predictions.transform.dense.bias       | UNEXPECTED |  | 
cls.seq_relationship.bias                  | UNEXPECTED |  | 
cls.seq_relationship.weight                | UNEXPECTED |  | 
cls.predictions.bias                       | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.
[transformers] loading configuration file config.json from cache at /root/.cache/huggingface/hub/models--bert-base-uncased/snapshots/86b5e0934494bd15c9632b12f734a8a67f723594/config.json
[transformers] Model config BertConfig {
  "add_cross_attention": false,
 

Loading weights:   0%|          | 0/202 [00:00<?, ?it/s]

[transformers] BertForMaskedLM LOAD REPORT from: bert-base-uncased
Key                         | Status     |  | 
----------------------------+------------+--+-
cls.seq_relationship.weight | UNEXPECTED |  | 
bert.pooler.dense.bias      | UNEXPECTED |  | 
bert.pooler.dense.weight    | UNEXPECTED |  | 
cls.seq_relationship.bias   | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.


In [ ]:
bert

BertModel(
  (embeddings): BertEmbeddings(
    (word_embeddings): Embedding(30522, 768, padding_idx=0)
    (position_embeddings): Embedding(512, 768)
    (token_type_embeddings): Embedding(2, 768)
    (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
    (dropout): Dropout(p=0.1, inplace=False)
  )
  (encoder): BertEncoder(
    (layer): ModuleList(
      (0-11): 12 x BertLayer(
        (attention): BertAttention(
          (self): BertSelfAttention(
            (query): Linear(in_features=768, out_features=768, bias=True)
            (key): Linear(in_features=768, out_features=768, bias=True)
            (value): Linear(in_features=768, out_features=768, bias=True)
            (dropout): Dropout(p=0.1, inplace=False)
          )
          (output): BertSelfOutput(
            (dense): Linear(in_features=768, out_features=768, bias=True)
            (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
            (dropout): Dropout(p=0.1, inplace=False)
  

In [ ]:
mlm

BertForMaskedLM(
  (bert): BertModel(
    (embeddings): BertEmbeddings(
      (word_embeddings): Embedding(30522, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (token_type_embeddings): Embedding(2, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (encoder): BertEncoder(
      (layer): ModuleList(
        (0-11): 12 x BertLayer(
          (attention): BertAttention(
            (self): BertSelfAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
            (output): BertSelfOutput(
              (dense): Linear(in_features=768, out_features=768, bias=True)
              (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_a

In [ ]:
text = ("After Abraham Lincoln won the November 1860 presidential "
        "election on an anti-slavery platform, an initial seven "
        "slave states declared their secession from the country "
        "to form the Confederacy. War broke out in April 1861 "
        "when secessionist forces attacked Fort Sumter in South "
        "Carolina, just over a month after Lincoln's "
        "inauguration.")

In [ ]:
text

"After Abraham Lincoln won the November 1860 presidential election on an anti-slavery platform, an initial seven slave states declared their secession from the country to form the Confederacy. War broke out in April 1861 when secessionist forces attacked Fort Sumter in South Carolina, just over a month after Lincoln's inauguration."

In [ ]:
inputs = tokenizer(text,return_tensors='pt')

In [ ]:
inputs

{'input_ids': tensor([[  101,  2044,  8181,  5367,  2180,  1996,  2281,  7313,  4883,  2602,
          2006,  2019,  3424,  1011,  8864,  4132,  1010,  2019,  3988,  2698,
          6658,  2163,  4161,  2037, 22965,  2013,  1996,  2406,  2000,  2433,
          1996, 18179,  1012,  2162,  3631,  2041,  1999,  2258,  6863,  2043,
         22965,  2923,  2749,  4457,  3481,  7680,  3334,  1999,  2148,  3792,
          1010,  2074,  2058,  1037,  3204,  2044,  5367,  1005,  1055, 17331,
          1012,   102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])}

In [ ]:
inputs['input_ids'].shape

torch.Size([1, 62])

inputs['input_ids'][0]：提取单条数据的 ID
在上一行代码中，你打印了 inputs['input_ids'].shape，结果是 [1, 62]。这表示当前的数据是一个批次（Batch），批次大小为 1，包含了 62 个词元的 ID。

加上 [0] 相当于进行切片降维，把外层代表 Batch 的那个维度去掉，仅仅提取出这 62 个数字组成的一维数组（Tensor）。

tokenizer.convert_ids_to_tokens(...)：数字转词元
这是 Hugging Face 分词器自带的一个极其常用的方法。

模型是不认识英文字母的，它只认识数字（比如 101, 2023 等）。这个方法相当于去查字典（Vocabulary），把这 62 个数字 ID 一一映射回对应的英文词块（Token）列表。

在这个阶段，输出的其实是一个 Python 列表（List），里面全是单个的字符串，比如 ['[CLS]', 'after', 'abraham', ...]。

' '.join(...)：拼接成完整字符串
这是 Python 原生的字符串操作方法。

它把上一步得到的包含 62 个字符串的列表，用空格（' '）作为分隔符，首尾相连拼成一个完整的长字符串，最终打印在屏幕上。

In [ ]:
' '.join(tokenizer.convert_ids_to_tokens(inputs['input_ids'][0]))

"[CLS] after abraham lincoln won the november 1860 presidential election on an anti - slavery platform , an initial seven slave states declared their secession from the country to form the confederacy . war broke out in april 1861 when secession ##ist forces attacked fort sum ##ter in south carolina , just over a month after lincoln ' s inauguration . [SEP]"

# masking

In [ ]:
inputs['labels'] = inputs['input_ids'].detach().clone()

In [ ]:
inputs['labels']

tensor([[  101,  2044,  8181,  5367,  2180,  1996,  2281,  7313,  4883,  2602,
          2006,  2019,  3424,  1011,  8864,  4132,  1010,  2019,  3988,  2698,
          6658,  2163,  4161,  2037, 22965,  2013,  1996,  2406,  2000,  2433,
          1996, 18179,  1012,  2162,  3631,  2041,  1999,  2258,  6863,  2043,
         22965,  2923,  2749,  4457,  3481,  7680,  3334,  1999,  2148,  3792,
          1010,  2074,  2058,  1037,  3204,  2044,  5367,  1005,  1055, 17331,
          1012,   102]])

In [ ]:
mask = torch.rand(inputs['input_ids'].shape) < 0.15

In [ ]:
mask

tensor([[ True, False,  True, False,  True, False, False, False, False, False,
          True, False,  True, False, False, False, False, False, False, False,
         False, False, False, False, False, False, False, False, False, False,
          True,  True, False, False,  True, False,  True, False, False, False,
         False, False, False, False, False, False, False, False, False, False,
         False, False, False, False, False, False, False, False, False,  True,
         False, False]])

In [ ]:
sum(mask[0])

tensor(10)

In [ ]:
10/62

0.16129032258064516

这里取出了第一句话的掩码数组 mask[0]，并使用 sum() 计算里面有多少个 True。

结果算出来有 10 个词元被挑中了。

接着做了一个简单的数学除法 10 / 62。结果是 16.1%，这非常接近我们设定的 15% 的目标概率，证明随机生成是符合预期的。
这行代码就是为了解决这个问题：

条件 1：(torch.rand(...) < 0.15)
和前面一样，先保证 15% 的基础随机选中率。

条件 2：(inputs['input_ids'] != 101)
在 BERT (基于 bert-base-uncased) 的词表中，数字 101 代表 [CLS]。这个条件保证了当前词元不是 [CLS] 时，结果才为 True。

条件 3：(inputs['input_ids'] != 102)
同理，数字 102 代表 [SEP]。这个条件保证了当前词元不是 [SEP] 时，结果才为 True。

* (乘法符号)：
在 PyTorch 中，当两个（或多个）布尔类型的张量进行相乘时，它代表的是逻辑与（AND）操作。

整段逻辑翻译成白话就是：

必须同时满足三个条件，一个词元才会被标记为 True（被遮掩）：

恰好命中了 15% 的随机概率。 并且

它的 ID 不是 101（不是 [CLS]）。 并且

它的 ID 不是 102（不是 [SEP]）。

In [ ]:
mask_arr = (torch.rand(inputs['input_ids'].shape)< 0.15) \
      * (inputs['input_ids'] != 101) \
      * (inputs['input_ids'] != 102)

In [ ]:
mask_arr

tensor([[False, False, False, False, False, False, False, False, False, False,
         False,  True, False, False,  True, False,  True,  True,  True, False,
         False, False, False, False, False, False, False,  True, False, False,
         False, False, False, False, False, False, False, False, False, False,
         False,  True, False, False,  True, False,  True, False, False, False,
         False, False, False, False, False,  True, False, False, False, False,
         False, False]])

In [ ]:
sum(mask_arr[0])

tensor(10)

mask_arr[0]：取出第一句话的布尔掩码数组（里面有 10 个 True，其余是 False）。

.nonzero()：这是一个非常实用的函数，它会找出数组中所有非零元素（在布尔数组中，True 就是 1，非零）的索引位置。

torch.flatten() 和 .tolist()：将提取出来的张量索引展平为一维，并转换成标准的 Python 列表。

结果：列表里的这 10 个数字，准确地指出了上一条代码中随机选中的 10 个词汇在这个句子里的具体位置（比如第 11 个词、第 14 个词等）。

In [ ]:
selection = torch.flatten(mask_arr[0].nonzero()).tolist()

In [ ]:
selection

[11, 14, 16, 17, 18, 27, 41, 44, 46, 55]

In [ ]:
tokenizer.special_tokens_map

{'unk_token': '[UNK]',
 'sep_token': '[SEP]',
 'pad_token': '[PAD]',
 'cls_token': '[CLS]',
 'mask_token': '[MASK]'}

这两行代码我们在上一个问题中已经接触过。它首先查阅字典，确认了 [MASK] 这个特殊符号在模型里的数字代号是 103。

随后，它定位到之前随机选出的 15% 的位置（也就是 selection 列表里的索引），把 input_ids 里原本的真实单词 ID 强行替换成了 103。

注意打印出的 inputs 字典：你会发现在 input_ids 这个张量里，确实多出了很多 103。但是，在 labels 这个张量里，对应位置依然保留着最原始的单词 ID。

In [ ]:
tokenizer.vocab['[MASK]']

103

In [ ]:
inputs['input_ids'][0,selection] = 103

In [ ]:
inputs

{'input_ids': tensor([[  101,  2044,  8181,  5367,  2180,  1996,  2281,  7313,  4883,  2602,
          2006,   103,  3424,  1011,   103,  4132,   103,   103,   103,  2698,
          6658,  2163,  4161,  2037, 22965,  2013,  1996,   103,  2000,  2433,
          1996, 18179,  1012,  2162,  3631,  2041,  1999,  2258,  6863,  2043,
         22965,   103,  2749,  4457,   103,  7680,   103,  1999,  2148,  3792,
          1010,  2074,  2058,  1037,  3204,   103,  5367,  1005,  1055, 17331,
          1012,   102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]), 'labels': tensor([

把打了马赛克的 input_ids 重新转换回文本。再重新将答案对照组转换为正确文本对比

In [ ]:
' '.join(tokenizer.convert_ids_to_tokens(inputs['input_ids'][0]))

"[CLS] after abraham lincoln won the november 1860 presidential election on [MASK] anti - [MASK] platform [MASK] [MASK] [MASK] seven slave states declared their secession from the [MASK] to form the confederacy . war broke out in april 1861 when secession [MASK] forces attacked [MASK] sum [MASK] in south carolina , just over a month [MASK] lincoln ' s inauguration . [SEP]"

In [ ]:
' '.join(tokenizer.convert_ids_to_tokens(inputs['labels'][0]))

"[CLS] after abraham lincoln won the november 1860 presidential election on an anti - slavery platform , an initial seven slave states declared their secession from the country to form the confederacy . war broke out in april 1861 when secession ##ist forces attacked fort sum ##ter in south carolina , just over a month after lincoln ' s inauguration . [SEP]"

# forward and calculate loss

In [ ]:
mlm.eval()
with torch.no_grad():
  output = mlm(**inputs)

In [ ]:
output.keys()

odict_keys(['loss', 'logits', 'hidden_states'])

In [ ]:
output.logits

tensor([[[ -7.3291,  -7.2502,  -7.3095,  ...,  -6.4581,  -6.4373,  -4.5043],
         [-10.8442, -10.8040, -10.8891,  ..., -10.5346,  -9.8068,  -8.6699],
         [ -6.4486,  -6.6094,  -6.0328,  ...,  -6.4274,  -6.3637,  -5.4772],
         ...,
         [ -2.3640,  -2.2651,  -2.2844,  ...,  -1.5560,  -1.3206,  -7.2121],
         [-13.8984, -13.9734, -14.0526,  ..., -11.1192, -10.7374,  -9.0886],
         [-12.0612, -12.4782, -12.1742,  ..., -11.7761, -10.1089,  -8.6747]]])

In [ ]:
output.loss

tensor(1.0567)

In [ ]:
output['hidden_states'] [-1]

tensor([[[-0.4119,  0.1023, -0.4152,  ..., -0.4361, -0.0815,  0.3993],
         [-0.5054, -0.0820,  0.0218,  ..., -0.4077,  0.2866,  0.4839],
         [-0.4408,  0.9003, -0.6498,  ..., -0.6054, -0.3316,  0.4066],
         ...,
         [-0.6157,  0.8164, -0.4936,  ..., -0.9678, -0.5554,  1.2254],
         [ 0.5237, -0.0206, -0.3793,  ...,  0.0796, -0.5075, -0.0271],
         [-0.1353, -0.0277, -0.7512,  ..., -0.3445, -0.6089,  0.1521]]])

# from scratch

In [ ]:
mlm.cls(output['hidden_states'][-1])

tensor([[[ -7.3291,  -7.2502,  -7.3095,  ...,  -6.4581,  -6.4373,  -4.5043],
         [-10.8442, -10.8040, -10.8891,  ..., -10.5346,  -9.8068,  -8.6699],
         [ -6.4486,  -6.6094,  -6.0328,  ...,  -6.4274,  -6.3637,  -5.4772],
         ...,
         [ -2.3640,  -2.2651,  -2.2844,  ...,  -1.5560,  -1.3206,  -7.2121],
         [-13.8984, -13.9734, -14.0526,  ..., -11.1192, -10.7374,  -9.0886],
         [-12.0612, -12.4782, -12.1742,  ..., -11.7761, -10.1089,  -8.6747]]],
       grad_fn=<ViewBackward0>)

In [ ]:
output.logits

tensor([[[ -7.3291,  -7.2502,  -7.3095,  ...,  -6.4581,  -6.4373,  -4.5043],
         [-10.8442, -10.8040, -10.8891,  ..., -10.5346,  -9.8068,  -8.6699],
         [ -6.4486,  -6.6094,  -6.0328,  ...,  -6.4274,  -6.3637,  -5.4772],
         ...,
         [ -2.3640,  -2.2651,  -2.2844,  ...,  -1.5560,  -1.3206,  -7.2121],
         [-13.8984, -13.9734, -14.0526,  ..., -11.1192, -10.7374,  -9.0886],
         [-12.0612, -12.4782, -12.1742,  ..., -11.7761, -10.1089,  -8.6747]]])

In [ ]:
last_hidden_state = output['hidden_states'][-1]

In [ ]:
mlm.eval()
with torch.no_grad():
  transformed = mlm.cls.predictions.transform(last_hidden_state)
  print(transformed.shape)
  logits = mlm.cls.predictions.decoder(transformed)
  print(logits.shape)
logits

torch.Size([1, 62, 768])
torch.Size([1, 62, 30522])


tensor([[[ -7.3291,  -7.2502,  -7.3095,  ...,  -6.4581,  -6.4373,  -4.5043],
         [-10.8442, -10.8040, -10.8891,  ..., -10.5346,  -9.8068,  -8.6699],
         [ -6.4486,  -6.6094,  -6.0328,  ...,  -6.4274,  -6.3637,  -5.4772],
         ...,
         [ -2.3640,  -2.2651,  -2.2844,  ...,  -1.5560,  -1.3206,  -7.2121],
         [-13.8984, -13.9734, -14.0526,  ..., -11.1192, -10.7374,  -9.0886],
         [-12.0612, -12.4782, -12.1742,  ..., -11.7761, -10.1089,  -8.6747]]])

In [ ]:
output.loss

tensor(1.0567)

In [ ]:
ce = nn.CrossEntropyLoss()

In [ ]:
logits.shape

torch.Size([1, 62, 30522])

In [ ]:
inputs['labels'].shape

torch.Size([1, 62])

In [ ]:
inputs['labels'][0].view(-1).shape

torch.Size([62])

In [ ]:
ce(logits[0],inputs['labels'][0].view(-1))

tensor(1.0567)

In [ ]:
torch.argmax(logits[0],dim=1)

tensor([ 1012,  2044,  8181,  5367,  2180,  1996,  2281,  7313,  4883,  2602,
         2006,  2019,  3424,  1011,  8864,  4132,  1010,  2035,  1996,  2698,
         6658,  2163,  4161,  2037, 22965,  2013,  1996,  2586,  2000,  2433,
         1996, 18179,  1012,  2162,  3631,  2041,  1999,  2258,  6863,  2043,
        22965,  2923,  2749,  4457,  3481,  7680,  3334,  1999,  2148,  3792,
         1010,  2074,  2058,  1037,  3204,  2044,  5367,  1005,  1055, 17331,
         1012,  1055])

In [ ]:
' '.join(tokenizer.convert_ids_to_tokens(inputs['input_ids'][0]))

"[CLS] after abraham lincoln won the november 1860 presidential election on [MASK] anti - [MASK] platform [MASK] [MASK] [MASK] seven slave states declared their secession from the [MASK] to form the confederacy . war broke out in april 1861 when secession [MASK] forces attacked [MASK] sum [MASK] in south carolina , just over a month [MASK] lincoln ' s inauguration . [SEP]"

In [ ]:
' '.join(tokenizer.convert_ids_to_tokens(torch.argmax(logits[0],dim=1)))

". after abraham lincoln won the november 1860 presidential election on an anti - slavery platform , all the seven slave states declared their secession from the union to form the confederacy . war broke out in april 1861 when secession ##ist forces attacked fort sum ##ter in south carolina , just over a month after lincoln ' s inauguration . s"

In [ ]:
' '.join(tokenizer.convert_ids_to_tokens(inputs['labels'][0]))

"[CLS] after abraham lincoln won the november 1860 presidential election on an anti - slavery platform , an initial seven slave states declared their secession from the country to form the confederacy . war broke out in april 1861 when secession ##ist forces attacked fort sum ##ter in south carolina , just over a month after lincoln ' s inauguration . [SEP]"